OpenRA.Mods.Common/Traits/World/Pathfinder/BasePathSearch.cs

```diff
@@ -0,0 +1,172 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using OpenRA.Primitives;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public interface IPathSearch
+	{
+		string Id { get; }
+		CellLayer<CellInfo> CellInfo { get; }
+		IPriorityQueue<CPos> OpenQueue { get; }
+		IEnumerable<Pair<CPos, int>> Considered { get; }
+		bool Debug { get; set; }
+		Player Owner { get; }
+		int MaxCost { get; }
+		IPathSearch Reverse();
+		IPathSearch WithCustomBlocker(Func<CPos, bool> customBlock);
+		IPathSearch WithIgnoredActor(Actor b);
+		IPathSearch WithHeuristic(Func<CPos, int> h);
+		IPathSearch WithCustomCost(Func<CPos, int> w);
+		IPathSearch WithoutLaneBias();
+		IPathSearch FromPoint(CPos from);
+		bool IsTarget(CPos location);
+		CPos Expand();
+	}
+
+	public abstract class BasePathSearch : IPathSearch
+	{
+		// The Id of a Pathsearch is computed by its properties.
+		// So two PathSearch instances with the same parameters will
+		// Compute the same Id. This is used for caching purposes.
+		public string Id
```

Comment:
Just pointing out that shortest paths are not actually the goal --
reasonable paths are.

This is starting to look pretty good though.

On Fri, Feb 13, 2015 at 12:15 PM, Paul Chote <notifications@github.com>
wrote:

> In OpenRA.Mods.Common/Traits/World/Pathfinder/BasePathSearch.cs
> <https://github.com/OpenRA/OpenRA/pull/7430#discussion_r24630876>:
>
> > +		IPathSearch WithCustomBlocker(Func<CPos, bool> customBlock);
> > +		IPathSearch WithIgnoredActor(Actor b);
> > +		IPathSearch WithHeuristic(Func<CPos, int> h);
> > +		IPathSearch WithCustomCost(Func<CPos, int> w);
> > +		IPathSearch WithoutLaneBias();
> > +		IPathSearch FromPoint(CPos from);
> > +		bool IsTarget(CPos location);
> > +		CPos Expand();
> > +	}
> > +
> > +	public abstract class BasePathSearch : IPathSearch
> > +	{
> > +		// The Id of a Pathsearch is computed by its properties.
> > +		// So two PathSearch instances with the same parameters will
> > +		// Compute the same Id. This is used for caching purposes.
> > +		public string Id
>
> This looks suspiciously like a DIY GetHashCode(). Does it really need to
> be a string?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenRA/OpenRA/pull/7430/files#r24630876>.
>